profiling/src/profile/pprof/sliced_proto.rs

```diff
@@ -0,0 +1,153 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present Datadog, Inc.
+
+//! This file is a companion to the standard pprof protobuf definition.
+//!
+//! When `prost` serializes the standard `pprof` protobuf, it does a batch
+//! operation:
+//! 1. the entire `internal::Profile` is converted to `pprof::Profile`,
+//! 2. which is then serialized into a byte-vector holding the protobuf,
+//! 3. which is then compressed.
+//!
+//! This operation is memory inefficient:
+//! 1. `pprof::Profile` does not deduplicate stack traces or labels, causing
+//!    significant memory blow-up compared to `internal::Profile` for timelined
+//!    traces.
+//! 2. The `pprof` protobuf is highly compressible, particularly when timeline
+//!    is enabled.  We are storing in memory a large buffer, that could have
+//!    easily been compressed to a small one.  
+//! 3. Only at this point do we have a small in-memory buffer.
+//!
+//! If we stream the creation of the protobuf, we can avoid this memory blowup.
+//! In particular, if each `pprof` Message (e.g. Sample, Location, etc) is
+//! generated in a streaming fashion, and then immediately serialized and
+//! compressed, we go directly from the efficient `internal` format to an
+//! a compact compressed serialized format, without needing to ever create large
+//! intermediate data-structures.
+//!
+//! We do this by taking advantage of the fact that the top level message in a
+//! protobuf doesn't have a length header.  This means that the bits on the wire
+//! are indistinguishable between serializing a single the top-level message,
```

Comment:
Minor:

```suggestion
//! are indistinguishable between serializing a single top-level message,
```

Also the paragraph mixes "top-level" and "top level" ;)